java-connectors/kafka-connect-azure-eventhubs/src/main/java/io/lenses/java/streamreactor/connect/azure/eventhubs/source/EventHubsKafkaConsumerController.java

```diff
@@ -0,0 +1,144 @@
+    this.recordsQueue = recordsQueue;
+
+    queuedKafkaConsumer = createQueuedConsumer();
+  }
+
+  private BlockingQueuedKafkaConsumer createQueuedConsumer() {
+    Properties consumerProperties = extractConsumerProperties();
+    return kafkaConsumerProvider.createConsumer(consumerProperties, recordsQueue);
+  }
+
+  private Properties extractConsumerProperties() {
+    Properties consumerProperties = new Properties();
+
+    consumerProperties.put(BOOTSTRAP_SERVERS_CONFIG,
+        azureEventHubsConfig.getList(
+            getPrefixedKafkaConsumerConfigKey(BOOTSTRAP_SERVERS_CONFIG)));
+    consumerProperties.put(GROUP_ID_CONFIG,
+        azureEventHubsConfig.getString(
+            getPrefixedKafkaConsumerConfigKey(GROUP_ID_CONFIG)));
+    consumerProperties.put(SaslConfigs.SASL_JAAS_CONFIG,
+        azureEventHubsConfig.getPassword(
+            getPrefixedKafkaConsumerConfigKey(SaslConfigs.SASL_JAAS_CONFIG)));
+    consumerProperties.put(SaslConfigs.SASL_MECHANISM,
+        azureEventHubsConfig.getString(
+            getPrefixedKafkaConsumerConfigKey(SaslConfigs.SASL_MECHANISM)));
+    consumerProperties.put(CommonClientConfigs.SECURITY_PROTOCOL_CONFIG,
+        azureEventHubsConfig.getString(
+            getPrefixedKafkaConsumerConfigKey(CommonClientConfigs.SECURITY_PROTOCOL_CONFIG)));
+    consumerProperties.put(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG,
+        azureEventHubsConfig.getClass(
+            getPrefixedKafkaConsumerConfigKey(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG)));
+    consumerProperties.put(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG,
+        azureEventHubsConfig.getClass(
+            getPrefixedKafkaConsumerConfigKey(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG)));
+    consumerProperties.put(ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG, "false");
+
+    return consumerProperties;
+  }
+
+  /**
+   * This method leverages BlockingQueue mechanism that BlockingQueuedKafkaConsumer puts EventHub records
+   * into. It tries to poll the queue then returns list of SourceRecords
+   *
+   * @param duration how often to poll.
+   * @return list of SourceRecords (can be empty if it couldn't poll from queue)
+   * @throws InterruptedException if interrupted while polling
+   */
+  public List<SourceRecord> poll(Duration duration) throws InterruptedException {
+    boolean includeHeaders = azureEventHubsConfig.getBoolean(
+        AzureEventHubsConfigConstants.INCLUDE_HEADERS);
+    List<SourceRecord> sourceRecords = null;
+
+    queuedKafkaConsumer.startPolling(duration);
+
+    ConsumerRecords<String, String> consumerRecords = null;
+    try {
+      consumerRecords = recordsQueue.poll(
+          duration.get(ChronoUnit.SECONDS), TimeUnit.SECONDS);
+    } catch (InterruptedException e) {
+      log.info("{} has been interrupted on poll", this.getClass().getSimpleName());
+      throw e;
+    }
+
+    if (consumerRecords != null && !consumerRecords.isEmpty()) {
```

Comment:
The null and empty check can be avoided if there is a decision where the code never returns null. The the logic of the method can be reduced to:

```
if(!empty){
return consumerRecords.stream.map{
  convert to source record
} .collect(Collectors.toList).
} else{
return Collections.emptyList();
}
```

This also avoids checking on sourceRecord being initialised.

Reply:
Unfortunately this is not true: `poll()` method can (and will!) return null record if there's nothing in the Queue, so at least on this side of the queue we have to have this check.